pkg/snapstore/utils.go

```diff
@@ -76,18 +80,35 @@ func GetEnvVarOrError(varName string) (string, error) {
 }
 
 // collectChunkUploadError collects the error from all go routine to upload individual chunks
-func collectChunkUploadError(errCh chan chunkUploadError, noOfChunks int64) []chunkUploadError {
-	var snapshotErr []chunkUploadError
+func collectChunkUploadError(chunkUploadCh chan<- chunk, resCh <-chan chunkUploadResult, stopCh chan struct{}, noOfChunks int64) *chunkUploadResult {
 	remainingChunks := noOfChunks
 	logrus.Infof("No of Chunks:= %d", noOfChunks)
-	for {
-		chunkErr := <-errCh
-		if chunkErr.err != nil {
-			snapshotErr = append(snapshotErr, chunkErr)
+	for chunkRes := range resCh {
+		logrus.Infof("Received chunk result for id: %d, offset: %d", chunkRes.chunk.id, chunkRes.chunk.offset)
+		if chunkRes.err != nil {
+			if chunkRes.chunk.attempt == maxRetryAttempts {
+				logrus.Errorf("Received the chunk upload error from one of the workers. Sending stop signal to all workers.")
+				close(stopCh)
+				return &chunkRes
+			}
+			chunk := chunkRes.chunk
+			delayTime := (1 << chunk.attempt)
+			chunk.attempt++
+			logrus.Warnf("Will try to upload chunk id:%d, offset: %d at attempt %d  after %d seconds", chunk.id, chunk.offset, chunk.attempt, delayTime)
+			time.AfterFunc(time.Duration(delayTime)*time.Second, func() {
+				select {
+				case <-stopCh:
+					return
+				default:
+					chunkUploadCh <- *chunk
+				}
+			})
 		}
 		remainingChunks--
```

Comment:
remainingChunks should be reduced only on a successful upload, ryt? So can we have it in an `else`?